<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
  <head>
    <title></title>
    <meta charset="utf-8" />
    <meta name="author" content="Rockefeller University, Bioinformatics Resource Centre" />
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="metropolisCustom.css" type="text/css" />
    <link rel="stylesheet" href="metropolis-fontsCustom.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Getting tidy
<html>
<div style="float:left">

</div>
<hr color='#EB811B' size=1px width=796px>
</html>
### Rockefeller University, Bioinformatics Resource Centre
### <a href="http://rockefelleruniversity.github.io/Intro_To_R_1Day/" class="uri">http://rockefelleruniversity.github.io/Intro_To_R_1Day/</a>

---


---



# BRC - Bioinformatic Resource Center

We are your local friendly bioinformaticians. 

We want to help with your problematic data. New or Old. 

Drop by and say Hi! We are currently in the IT pavillion. 

# Introduction to Data Wrangling with Tidy
---
## A problem we face when dealing with data regularly.

Every dataset is different. Sometimes very different.

There are many ways to do things. Everyone has their favorite syntax. 

The issue:  
Many fundamental data processing functions exist in *Base R* and beyond. Sometimes they can be inconsistent or unnecessarily complex. Espeically when dealing with non-standard dataframes. The result is code that is confusing and doesn't flow i.e. nested functions

---
## What does it mean to be tidy?

Tidyverse is most importantly a philosophy for data analysis that more often then not makes wrangling data easier. The tidyverse community have built what they describe as an "opinionated" group of packages. These packages readily talk to one another. 

* More efficient code
* Easier to remember syntax
* Easier to read syntax

Their manifesto: https://cran.r-project.org/web/packages/tidyverse/vignettes/manifesto.html

---
## What does it actually mean to be tidy?

* A defined vision for coding style in R  
* A defined vision for data formats in R
* A defined vision for package design in R
* Unified set of community pushing in a cohesive direction
* Critical mass of people to influence the way the whole R community evolves

---
## What are the main tools in the tidyverse?

* ggplot2 – making pretty graphs 
* readr – reading data into R  
* dplyr – manipulating data  
* tibble - working with tibbles  
* tidyr – miscellaneous tools for tidying data
* purrr - iterating over data  
* stringr – working with strings  
* forcats - working with factors  

Other tools have now been made for the tidy community. This community also overlaps with bioconductor. But the packages above are the linchpins that hold it together. 

---
## What are we doing today

Workflow Image for working with data.

---
## What we won't be doing today

We already have a course online for plotting, including ggplot.  
https://rockefelleruniversity.github.io/Plotting_In_R/

# Lets get tidy!
---
## First step lets load in the data we are using today

```r
cat(getwd())
```

```
## /Users/tcarroll/Projects/Software/teaching/tidyR
```

```r
load(file='dataset/my_tidy.Rdata')
```

---
## Are all data frames equal?

```r
head(df1)
```

```
## # A tibble: 6 x 5
##   salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml
##       &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;
## 1     35032 Chinook salmon yearling                 147       41.3
## 2     35035 Sockeye salmon juvenile                 121       NA  
## 3     35036 Sockeye salmon juvenile                 112       NA  
## 4     35037 Steelhead      juvenile                 220       42.7
## 5     35038 Steelhead      juvenile                 152       NA  
## 6     35033 Chinook salmon mixed age juvenile       444       62.1
```

```r
head(df2)
```

```
## # A tibble: 6 x 5
##   salmon_id common_name    age_classbylength  variable   value
##       &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;              &lt;fct&gt;      &lt;dbl&gt;
## 1     35032 Chinook salmon yearling           length_mm  147  
## 2     35032 Chinook salmon yearling           IGF1_ng_ml  41.3
## 3     35033 Chinook salmon mixed age juvenile length_mm  444  
## 4     35033 Chinook salmon mixed age juvenile IGF1_ng_ml  62.1
## 5     35034 Sockeye salmon juvenile           length_mm  139  
## 6     35034 Sockeye salmon juvenile           IGF1_ng_ml  NA
```

```r
head(df3a)
```

```
## # A tibble: 6 x 5
##   salmon_id common_name    age_classbylength  variable  value
##       &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;              &lt;fct&gt;     &lt;dbl&gt;
## 1     35032 Chinook salmon yearling           length_mm   147
## 2     35033 Chinook salmon mixed age juvenile length_mm   444
## 3     35034 Sockeye salmon juvenile           length_mm   139
## 4     35035 Sockeye salmon juvenile           length_mm   121
## 5     35036 Sockeye salmon juvenile           length_mm   112
## 6     35037 Steelhead      juvenile           length_mm   220
```

```r
head(df3b)
```

```
## # A tibble: 6 x 5
##   salmon_id common_name    age_classbylength  variable   value
##       &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;              &lt;fct&gt;      &lt;dbl&gt;
## 1     35032 Chinook salmon yearling           IGF1_ng_ml  41.3
## 2     35033 Chinook salmon mixed age juvenile IGF1_ng_ml  62.1
## 3     35034 Sockeye salmon juvenile           IGF1_ng_ml  NA  
## 4     35035 Sockeye salmon juvenile           IGF1_ng_ml  NA  
## 5     35036 Sockeye salmon juvenile           IGF1_ng_ml  NA  
## 6     35037 Steelhead      juvenile           IGF1_ng_ml  42.7
```

---
## What is a tidy dataset?

A tidy dataset is a data frame (or table) for which the following are true:

* Each variable has its own column
* Each observation has its own row
* Each value has its own cell

Our first dataframe is tidy

---
## Why bother?

Consistent dataframe layouts help to ensure that all values are present and that relationships between data points are clear.

R is a vectorized programming language. R builds data frames from vectors, and R works best when its operation are vectorized. Tidy data utilizes of both of these aspects of R.

=&gt; Precise and Fast

---
## Lets load in the tidyverse

```r
library(tidyverse)
```

```
## ── Attaching packages ────────────────────────────────── tidyverse 1.2.1 ──
```

```
## ✔ ggplot2 3.1.1     ✔ purrr   0.3.2
## ✔ tibble  2.1.1     ✔ dplyr   0.8.3
## ✔ tidyr   1.0.0     ✔ stringr 1.4.0
## ✔ readr   1.3.1     ✔ forcats 0.4.0
```

```
## ── Conflicts ───────────────────────────────────── tidyverse_conflicts() ──
## ✖ dplyr::filter() masks stats::filter()
## ✖ dplyr::lag()    masks stats::lag()
```

---
## These tools have same logic

 -&gt; insert graphic here
 https://github.com/trinker/tidyr_in_a_nutshell

# dplyr: A tool to access and manipulate data in a dataframe

---
## Select
Select allows you to make a vector or dataframe from a specific variable or variables

```r
# Select one variable (common_name)
select(df1, common_name)
```

```
## # A tibble: 97 x 1
##    common_name   
##    &lt;chr&gt;         
##  1 Chinook salmon
##  2 Sockeye salmon
##  3 Sockeye salmon
##  4 Steelhead     
##  5 Steelhead     
##  6 Chinook salmon
##  7 Sockeye salmon
##  8 Steelhead     
##  9 Steelhead     
## 10 Steelhead     
## # … with 87 more rows
```

```r
# Select two variables (age_classbylength and common_name)
select(df1, age_classbylength, common_name)
```

```
## # A tibble: 97 x 2
##    age_classbylength  common_name   
##    &lt;chr&gt;              &lt;chr&gt;         
##  1 yearling           Chinook salmon
##  2 juvenile           Sockeye salmon
##  3 juvenile           Sockeye salmon
##  4 juvenile           Steelhead     
##  5 juvenile           Steelhead     
##  6 mixed age juvenile Chinook salmon
##  7 juvenile           Sockeye salmon
##  8 juvenile           Steelhead     
##  9 juvenile           Steelhead     
## 10 juvenile           Steelhead     
## # … with 87 more rows
```

```r
# Select all but one variable (length_mm)
select(df1,-length_mm)
```

```
## # A tibble: 97 x 4
##    salmon_id common_name    age_classbylength  IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                   &lt;dbl&gt;
##  1     35032 Chinook salmon yearling                 41.3
##  2     35035 Sockeye salmon juvenile                 NA  
##  3     35036 Sockeye salmon juvenile                 NA  
##  4     35037 Steelhead      juvenile                 42.7
##  5     35038 Steelhead      juvenile                 NA  
##  6     35033 Chinook salmon mixed age juvenile       62.1
##  7     35034 Sockeye salmon juvenile                 NA  
##  8     35048 Steelhead      juvenile                 24.2
##  9     35049 Steelhead      juvenile                 NA  
## 10     35050 Steelhead      juvenile                 63.5
## # … with 87 more rows
```

```r
# Select all a range of contiguous varibles (common_name:length_mm)
select(df1, common_name:length_mm)
```

```
## # A tibble: 97 x 3
##    common_name    age_classbylength  length_mm
##    &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;
##  1 Chinook salmon yearling                 147
##  2 Sockeye salmon juvenile                 121
##  3 Sockeye salmon juvenile                 112
##  4 Steelhead      juvenile                 220
##  5 Steelhead      juvenile                 152
##  6 Chinook salmon mixed age juvenile       444
##  7 Sockeye salmon juvenile                 139
##  8 Steelhead      juvenile                 288
##  9 Steelhead      juvenile                 190
## 10 Steelhead      juvenile                 283
## # … with 87 more rows
```

---
## Filter
Filter allows you to access observations based on specific criteria

```r
# Filter all observation where the variable common_name is Sockeye salmon
filter(df1, common_name == 'Sockeye salmon')
```

```
## # A tibble: 11 x 5
##    salmon_id common_name    age_classbylength length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                 &lt;dbl&gt;      &lt;dbl&gt;
##  1     35035 Sockeye salmon juvenile                121         NA
##  2     35036 Sockeye salmon juvenile                112         NA
##  3     35034 Sockeye salmon juvenile                139         NA
##  4     35144 Sockeye salmon juvenile                140         NA
##  5     35147 Sockeye salmon juvenile                115         NA
##  6     35096 Sockeye salmon juvenile                115         NA
##  7     35097 Sockeye salmon juvenile                110         NA
##  8     35098 Sockeye salmon juvenile                112         NA
##  9     35099 Sockeye salmon juvenile                111         NA
## 10     35100 Sockeye salmon juvenile                118         NA
## 11     35119 Sockeye salmon juvenile                122         NA
```

```r
# Filter all observations where the variable common_name is either Sockeye salmon or Chinook Salmon
filter(df1, common_name %in% c('Sockeye salmon', 'Chinook salmon'))
```

```
## # A tibble: 57 x 5
##    salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;
##  1     35032 Chinook salmon yearling                 147       41.3
##  2     35035 Sockeye salmon juvenile                 121       NA  
##  3     35036 Sockeye salmon juvenile                 112       NA  
##  4     35033 Chinook salmon mixed age juvenile       444       62.1
##  5     35034 Sockeye salmon juvenile                 139       NA  
##  6     35142 Chinook salmon yearling                 149       66.5
##  7     35143 Chinook salmon yearling                 204       80.9
##  8     35144 Sockeye salmon juvenile                 140       NA  
##  9     35145 Chinook salmon yearling                 130       23.4
## 10     35146 Chinook salmon mixed age juvenile       422      101. 
## # … with 47 more rows
```

```r
# Filter all observations where the variable common_name ends with 'salmon'. To do this we use stringr function str_ends recognise strings that end with 'salmon'.
filter(df1, str_ends(common_name, 'salmon'))
```

```
## # A tibble: 59 x 5
##    salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;
##  1     35032 Chinook salmon yearling                 147       41.3
##  2     35035 Sockeye salmon juvenile                 121       NA  
##  3     35036 Sockeye salmon juvenile                 112       NA  
##  4     35033 Chinook salmon mixed age juvenile       444       62.1
##  5     35034 Sockeye salmon juvenile                 139       NA  
##  6     35142 Chinook salmon yearling                 149       66.5
##  7     35143 Chinook salmon yearling                 204       80.9
##  8     35144 Sockeye salmon juvenile                 140       NA  
##  9     35145 Chinook salmon yearling                 130       23.4
## 10     35146 Chinook salmon mixed age juvenile       422      101. 
## # … with 49 more rows
```

```r
# Filter all observations where the variable length_mm is greater than 200 or less than 120
filter(df1, length_mm &gt; 200 | length_mm &lt; 120)
```

```
## # A tibble: 36 x 5
##    salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;
##  1     35036 Sockeye salmon juvenile                 112       NA  
##  2     35037 Steelhead      juvenile                 220       42.7
##  3     35033 Chinook salmon mixed age juvenile       444       62.1
##  4     35048 Steelhead      juvenile                 288       24.2
##  5     35050 Steelhead      juvenile                 283       63.5
##  6     35051 Steelhead      juvenile                 279       61.2
##  7     35052 Steelhead      juvenile                 235       30.6
##  8     35053 Steelhead      juvenile                 230       49.4
##  9     35056 Steelhead      juvenile                 208       57.4
## 10     35057 Steelhead      juvenile                 240       20.2
## # … with 26 more rows
```

---
## CHALLENGE

* Create a dataframe of the variables *age* and *IGF* of only the *Steelhead* fish
* Create a dataframe of all variables but the *IGF* values, for all fish that begin with *S*

---
## ANSWERS

```r
# Create a dataframe of the variables *age* and *IGF* of only the *Steelhead* fish
df1_A &lt;- filter(df1, common_name == 'Steelhead')
select(df1_A, age_classbylength, IGF1_ng_ml)
```

```
## # A tibble: 38 x 2
##    age_classbylength IGF1_ng_ml
##    &lt;chr&gt;                  &lt;dbl&gt;
##  1 juvenile                42.7
##  2 juvenile                NA  
##  3 juvenile                24.2
##  4 juvenile                NA  
##  5 juvenile                63.5
##  6 juvenile                61.2
##  7 juvenile                30.6
##  8 juvenile                49.4
##  9 juvenile                56.3
## 10 juvenile                55.7
## # … with 28 more rows
```

```r
# Create a dataframe of all variables but the *IGF* values, for all fish that begin with *S*
df1_A &lt;- filter(df1, str_starts(common_name,'S'))
select(df1_A, -IGF1_ng_ml)
```

```
## # A tibble: 49 x 4
##    salmon_id common_name    age_classbylength length_mm
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                 &lt;dbl&gt;
##  1     35035 Sockeye salmon juvenile                121
##  2     35036 Sockeye salmon juvenile                112
##  3     35037 Steelhead      juvenile                220
##  4     35038 Steelhead      juvenile                152
##  5     35034 Sockeye salmon juvenile                139
##  6     35048 Steelhead      juvenile                288
##  7     35049 Steelhead      juvenile                190
##  8     35050 Steelhead      juvenile                283
##  9     35051 Steelhead      juvenile                279
## 10     35052 Steelhead      juvenile                235
## # … with 39 more rows
```

---
## Arrange
Arrange sorts the dataframe based on a specific variable or variables

```r
# Arrange the data based on the variable length_mm
arrange(df1, length_mm)
```

```
## # A tibble: 97 x 5
##    salmon_id common_name    age_classbylength length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                 &lt;dbl&gt;      &lt;dbl&gt;
##  1     35095 Chinook salmon subyearling              90         NA
##  2     35097 Sockeye salmon juvenile                110         NA
##  3     35099 Sockeye salmon juvenile                111         NA
##  4     35036 Sockeye salmon juvenile                112         NA
##  5     35098 Sockeye salmon juvenile                112         NA
##  6     35147 Sockeye salmon juvenile                115         NA
##  7     35096 Sockeye salmon juvenile                115         NA
##  8     35100 Sockeye salmon juvenile                118         NA
##  9     35035 Sockeye salmon juvenile                121         NA
## 10     35119 Sockeye salmon juvenile                122         NA
## # … with 87 more rows
```

```r
# Arrange the data first based on the variable common_name, then secondly based on length_mm in a descending order.
arrange(df1, common_name, desc(length_mm))
```

```
## # A tibble: 97 x 5
##    salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;
##  1     35033 Chinook salmon mixed age juvenile       444      62.1 
##  2     35146 Chinook salmon mixed age juvenile       422     101.  
##  3     35110 Chinook salmon mixed age juvenile       275      81.5 
##  4     35129 Chinook salmon yearling                 225      72.7 
##  5     35103 Chinook salmon yearling                 216      81.2 
##  6     35115 Chinook salmon yearling                 215      53.5 
##  7     35112 Chinook salmon yearling                 205      90.5 
##  8     35143 Chinook salmon yearling                 204      80.9 
##  9     35079 Chinook salmon yearling                 199      53.2 
## 10     35081 Chinook salmon yearling                 196       5.56
## # … with 87 more rows
```

---
## Mutate
Mutate creates a new variable based on some form of computation

```r
# A new variable is created based on the calculation of the z-score of the variable IGF1_ng_ml using scale()
mutate(df1, scale(IGF1_ng_ml))
```

```
## # A tibble: 97 x 6
##    salmon_id common_name age_classbyleng… length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;       &lt;chr&gt;                &lt;dbl&gt;      &lt;dbl&gt;
##  1     35032 Chinook sa… yearling               147       41.3
##  2     35035 Sockeye sa… juvenile               121       NA  
##  3     35036 Sockeye sa… juvenile               112       NA  
##  4     35037 Steelhead   juvenile               220       42.7
##  5     35038 Steelhead   juvenile               152       NA  
##  6     35033 Chinook sa… mixed age juven…       444       62.1
##  7     35034 Sockeye sa… juvenile               139       NA  
##  8     35048 Steelhead   juvenile               288       24.2
##  9     35049 Steelhead   juvenile               190       NA  
## 10     35050 Steelhead   juvenile               283       63.5
## # … with 87 more rows, and 1 more variable: `scale(IGF1_ng_ml)`[,1] &lt;dbl&gt;
```

```r
# A new variable is created called IGFngml_zscore, based on the caluclation of the z-score of the variable IGF1_ng_ml using scale()
mutate(df1, IGFngml_zscore = scale(IGF1_ng_ml))
```

```
## # A tibble: 97 x 6
##    salmon_id common_name age_classbyleng… length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;       &lt;chr&gt;                &lt;dbl&gt;      &lt;dbl&gt;
##  1     35032 Chinook sa… yearling               147       41.3
##  2     35035 Sockeye sa… juvenile               121       NA  
##  3     35036 Sockeye sa… juvenile               112       NA  
##  4     35037 Steelhead   juvenile               220       42.7
##  5     35038 Steelhead   juvenile               152       NA  
##  6     35033 Chinook sa… mixed age juven…       444       62.1
##  7     35034 Sockeye sa… juvenile               139       NA  
##  8     35048 Steelhead   juvenile               288       24.2
##  9     35049 Steelhead   juvenile               190       NA  
## 10     35050 Steelhead   juvenile               283       63.5
## # … with 87 more rows, and 1 more variable: IGFngml_zscore[,1] &lt;dbl&gt;
```

---
## CHALLENGE

* Create a dataframe with a new variable that contains the rank of the *length* variable. Arrange this new data frame by *IGF* (lowest to highest).
* Create a dataframe with a new variable that is *IGF*/*length*. Arrange by this new variable (highest to lowest).

---
## ANSWERS


```r
# Create a dataframe with a new variable that contains the rank of the *length* variable. Arrange this new data frame by *IGF* (lowest to highest).
df1_A &lt;- mutate(df1, length_rank=rank(length_mm))
arrange(df1_A, IGF1_ng_ml)
```

```
## # A tibble: 97 x 6
##    salmon_id common_name  age_classbyleng… length_mm IGF1_ng_ml length_rank
##        &lt;dbl&gt; &lt;chr&gt;        &lt;chr&gt;                &lt;dbl&gt;      &lt;dbl&gt;       &lt;dbl&gt;
##  1     35045 Steelhead    juvenile               177       4.46        49  
##  2     35081 Chinook sal… yearling               196       5.56        63.5
##  3     35062 Steelhead    juvenile               198       5.90        65.5
##  4     35138 Chinook sal… yearling               142       9.17        18  
##  5     35039 Steelhead    juvenile               201       9.92        70  
##  6     35135 Chinook sal… yearling               181      11.5         52.5
##  7     35113 Chinook sal… yearling               162      14.2         38  
##  8     35064 Steelhead    juvenile               225      16.0         80.5
##  9     35082 Chinook sal… yearling               150      17.1         29.5
## 10     35057 Steelhead    juvenile               240      20.2         86  
## # … with 87 more rows
```

```r
#Create a dataframe with a new variable that is *IGF*/*length*. Arrange by this new variable (highest to lowest).
df1_A &lt;- mutate(df1, IGF_perlength=IGF1_ng_ml/length_mm)
arrange(df1_A, -IGF_perlength)
```

```
## # A tibble: 97 x 6
##    salmon_id common_name age_classbyleng… length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;       &lt;chr&gt;                &lt;dbl&gt;      &lt;dbl&gt;
##  1     35055 Steelhead   juvenile               123       55.7
##  2     35087 Coho salmon yearling               164       73.6
##  3     35142 Chinook sa… yearling               149       66.5
##  4     35112 Chinook sa… yearling               205       90.5
##  5     35092 Chinook sa… yearling               192       81.8
##  6     35117 Chinook sa… yearling               147       61.2
##  7     35090 Chinook sa… yearling               167       67.3
##  8     35143 Chinook sa… yearling               204       80.9
##  9     35059 Steelhead   juvenile               185       73.1
## 10     35101 Steelhead   juvenile               167       65.4
## # … with 87 more rows, and 1 more variable: IGF_perlength &lt;dbl&gt;
```

---
## Summarize
Summarize applies aggregating or summary function to a group

```r
# First we define the common_name as a group. 
df1_byname &lt;- group_by(df1, common_name)
# Summarise is used to count over the grouped common_names
summarise(df1_byname, count = n())
```

```
## # A tibble: 4 x 2
##   common_name    count
##   &lt;chr&gt;          &lt;int&gt;
## 1 Chinook salmon    46
## 2 Coho salmon        2
## 3 Sockeye salmon    11
## 4 Steelhead         38
```

```r
# Summarise is used to calculate mean IGF1_ng_ml over the grouped common_names
summarise(df1_byname, IGF1_ng_ml_ave = mean(IGF1_ng_ml, na.rm = T))
```

```
## # A tibble: 4 x 2
##   common_name    IGF1_ng_ml_ave
##   &lt;chr&gt;                   &lt;dbl&gt;
## 1 Chinook salmon           46.8
## 2 Coho salmon              73.6
## 3 Sockeye salmon          NaN  
## 4 Steelhead                46.1
```

---
## Group
Grouping can also help ask questions with other functions

```r
# Filter obsevrations with the 2 smallest length_mm for each grouped common_names
filter(df1_byname, rank(length_mm) &lt;= 2)
```

```
## # A tibble: 8 x 5
## # Groups:   common_name [4]
##   salmon_id common_name    age_classbylength length_mm IGF1_ng_ml
##       &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                 &lt;dbl&gt;      &lt;dbl&gt;
## 1     35038 Steelhead      juvenile                152       NA  
## 2     35055 Steelhead      juvenile                123       55.7
## 3     35145 Chinook salmon yearling                130       23.4
## 4     35085 Coho salmon    yearling                140       NA  
## 5     35087 Coho salmon    yearling                164       73.6
## 6     35095 Chinook salmon subyearling              90       NA  
## 7     35097 Sockeye salmon juvenile                110       NA  
## 8     35099 Sockeye salmon juvenile                111       NA
```

```r
# Filter obsevrations with at least 5 for each grouped common_names
filter(df1_byname, n() &gt; 5)
```

```
## # A tibble: 95 x 5
## # Groups:   common_name [3]
##    salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;
##  1     35032 Chinook salmon yearling                 147       41.3
##  2     35035 Sockeye salmon juvenile                 121       NA  
##  3     35036 Sockeye salmon juvenile                 112       NA  
##  4     35037 Steelhead      juvenile                 220       42.7
##  5     35038 Steelhead      juvenile                 152       NA  
##  6     35033 Chinook salmon mixed age juvenile       444       62.1
##  7     35034 Sockeye salmon juvenile                 139       NA  
##  8     35048 Steelhead      juvenile                 288       24.2
##  9     35049 Steelhead      juvenile                 190       NA  
## 10     35050 Steelhead      juvenile                 283       63.5
## # … with 85 more rows
```

```r
# A new variable is created using z-score within the grouped common_names
mutate(df1_byname, IGFngml_zscore = scale(IGF1_ng_ml))
```

```
## # A tibble: 97 x 6
## # Groups:   common_name [4]
##    salmon_id common_name age_classbyleng… length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;       &lt;chr&gt;                &lt;dbl&gt;      &lt;dbl&gt;
##  1     35032 Chinook sa… yearling               147       41.3
##  2     35035 Sockeye sa… juvenile               121       NA  
##  3     35036 Sockeye sa… juvenile               112       NA  
##  4     35037 Steelhead   juvenile               220       42.7
##  5     35038 Steelhead   juvenile               152       NA  
##  6     35033 Chinook sa… mixed age juven…       444       62.1
##  7     35034 Sockeye sa… juvenile               139       NA  
##  8     35048 Steelhead   juvenile               288       24.2
##  9     35049 Steelhead   juvenile               190       NA  
## 10     35050 Steelhead   juvenile               283       63.5
## # … with 87 more rows, and 1 more variable: IGFngml_zscore &lt;dbl&gt;
```

---
## CHALLENGE

1. Group df1 by the variable *age_class*. 
2. Filter to get the biggest 5 by the variable *length* in each group. 
3. Summarise this data frame over the variable *length* by calculating the mean. 

---
## ANSWERS


```r
# Group df1 by the variable *age_class*. 
df1_A &lt;- group_by(df1, age_classbylength)

# Filter to get the biggest 5 by the variable *length* in each group. 
df1_A &lt;- filter(df1_A, rank(-length_mm) &lt;= 5)

# Summarise this data frame over the variable *length* by calculating the mean.
summarise(df1_A, mean_length_mm = mean(length_mm, na.rm = T))
```

```
## # A tibble: 4 x 2
##   age_classbylength  mean_length_mm
##   &lt;chr&gt;                       &lt;dbl&gt;
## 1 juvenile                     274 
## 2 mixed age juvenile           380.
## 3 subyearling                   90 
## 4 yearling                     213
```

# Piping from magrittR (%&gt;%): A way to string functions together
Piping allows you to pass the result from one expression directly into another.

 -&gt; same graphic as before , but extend
 https://github.com/trinker/tidyr_in_a_nutshell

---
## Piping versus not piping

```r
# Without pipe
df1_byname &lt;- group_by(df1, common_name)
summarise(df1_byname, IGF1_ng_ml_ave = mean(IGF1_ng_ml, na.rm=T))
```

```
## # A tibble: 4 x 2
##   common_name    IGF1_ng_ml_ave
##   &lt;chr&gt;                   &lt;dbl&gt;
## 1 Chinook salmon           46.8
## 2 Coho salmon              73.6
## 3 Sockeye salmon          NaN  
## 4 Steelhead                46.1
```

```r
# With pipe
group_by(df1, common_name) %&gt;% summarise(IGF1_ng_ml_ave = mean(IGF1_ng_ml, na.rm=T))
```

```
## # A tibble: 4 x 2
##   common_name    IGF1_ng_ml_ave
##   &lt;chr&gt;                   &lt;dbl&gt;
## 1 Chinook salmon           46.8
## 2 Coho salmon              73.6
## 3 Sockeye salmon          NaN  
## 4 Steelhead                46.1
```

---
## Building pipes together

```r
# (1) Group by common_name
# (2) Filter to all those that have length bigger then 200
# (3) Summarise is used to calculate mean IGF1_ng_ml over the grouped common_names for these larger fish
group_by(df1, common_name) %&gt;% 
  filter(length_mm &gt; 200) %&gt;% 
  summarise(IGF1_ng_ml_ave = mean(IGF1_ng_ml, na.rm = T))
```

```
## # A tibble: 2 x 2
##   common_name    IGF1_ng_ml_ave
##   &lt;chr&gt;                   &lt;dbl&gt;
## 1 Chinook salmon           77.9
## 2 Steelhead                45.3
```

```r
# (1) Create new variable that is discrete label depending on size of the fish
# (2) Group by common_name and size
# (3) Summarise is used to calculate mean IGF1_ng_ml over the grouped common_names and sizes
mutate(df1, size = if_else(length_mm &gt; 200, 'big_fish', 'small_fish')) %&gt;% 
  group_by(common_name, size) %&gt;% 
  summarise(IGF1_ng_ml_ave = mean(IGF1_ng_ml, na.rm = T))
```

```
## # A tibble: 6 x 3
## # Groups:   common_name [4]
##   common_name    size       IGF1_ng_ml_ave
##   &lt;chr&gt;          &lt;chr&gt;               &lt;dbl&gt;
## 1 Chinook salmon big_fish             77.9
## 2 Chinook salmon small_fish           39.5
## 3 Coho salmon    small_fish           73.6
## 4 Sockeye salmon small_fish          NaN  
## 5 Steelhead      big_fish             45.3
## 6 Steelhead      small_fish           47.3
```

```r
# (1) Create new variable that is discrete label depending on size of the fish
# (2) Group by common_name and size
# (3) Summarise is used to calculate mean IGF1_ng_ml over the grouped common_names and sizes
# (4) Filter out Coho and Sockeye salmon
mutate(df1, size = if_else(length_mm &gt; 200, 'big_fish', 'small_fish')) %&gt;% 
  group_by(common_name, size) %&gt;% 
  summarise(IGF1_ng_ml_ave = mean(IGF1_ng_ml, na.rm = T)) %&gt;%
  filter(common_name != 'Coho salmon')  %&gt;%
  filter(common_name !='Sockeye salmon')
```

```
## # A tibble: 4 x 3
## # Groups:   common_name [2]
##   common_name    size       IGF1_ng_ml_ave
##   &lt;chr&gt;          &lt;chr&gt;               &lt;dbl&gt;
## 1 Chinook salmon big_fish             77.9
## 2 Chinook salmon small_fish           39.5
## 3 Steelhead      big_fish             45.3
## 4 Steelhead      small_fish           47.3
```

---
## You can put these pipes straight into plotting

```r
p &lt;- mutate(df1,size=if_else(length_mm&gt;200, 'big_fish', 'small_fish')) %&gt;% 
  group_by(common_name, size) %&gt;% 
  summarize(IGF1_ng_ml_ave=mean(IGF1_ng_ml, na.rm=T)) %&gt;% 
  filter(common_name != 'Coho salmon')  %&gt;% 
  filter(common_name != 'Sockeye salmon') %&gt;% 
  ggplot( aes(x = common_name, y = IGF1_ng_ml_ave, group = size, fill = size)) +
  geom_bar(stat = "identity", position = "dodge") + 
  theme(axis.text.x = element_text(angle = 90)) +
  scale_fill_brewer(palette = "Paired")

p
```

![](20191009_tidyr_files/figure-html/unnamed-chunk-15-1.png)&lt;!-- --&gt;

---
## CHALLENGE

1. Filter to only look at *juvenile* in variable 
2. Group by common_name
3. Create new variable that is z-score of *length* across groups
4. Create boxplot of grouped length zscores

---
## ANSWER

```r
df1 %&gt;% 
  filter(age_classbylength == 'yearling') %&gt;%
  group_by(common_name) %&gt;% 
  mutate(length_zscore = scale(length_mm)) %&gt;%
  ggplot(aes(x = common_name, y = length_zscore)) +
  geom_boxplot(fill=c('gold','darkorange'))
```

![](20191009_tidyr_files/figure-html/unnamed-chunk-16-1.png)&lt;!-- --&gt;

# Readr: Reading data into R
So we blasted through what being tidy can give you. 
Now lets start from the beginning and tidy some data. First step is to read in data. 

ReadR:

* read_csv(): comma separated (CSV) files  
* read_tsv(): tab separated files  
* read_delim(): general delimited files  
* read_fwf(): fixed width files  
* read_table(): tabular files where columns are separated by white-space  
* read_log(): web log files   

---
## ReadR vs Base

```r
# Base gives you everything
untidy_counts_base &lt;- read.csv("dataset/hemato_rnaseq_counts.csv")
head(untidy_counts_base)
```

```
##   ENTREZ CD34_1 ORTHO_1 CD34_2 ORTHO_2
## 1    350    204       0    103       0
## 2    351  15586     479  10476      39
## 3    353    842     355   1188      86
## 4    354      0       0      0       0
## 5    355    123     291    139      16
## 6    356      1       1      0       0
```

```r
# ReadR gives you a tibble. 
# Note the rownames are now their own column.
untidy_counts &lt;- read_csv("dataset/hemato_rnaseq_counts.csv")
```

```
## Parsed with column specification:
## cols(
##   ENTREZ = col_double(),
##   CD34_1 = col_double(),
##   ORTHO_1 = col_double(),
##   CD34_2 = col_double(),
##   ORTHO_2 = col_double()
## )
```

```r
untidy_counts
```

```
## # A tibble: 100 x 5
##    ENTREZ CD34_1 ORTHO_1 CD34_2 ORTHO_2
##     &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;
##  1    350    204       0    103       0
##  2    351  15586     479  10476      39
##  3    353    842     355   1188      86
##  4    354      0       0      0       0
##  5    355    123     291    139      16
##  6    356      1       1      0       0
##  7    357    380       3    177       0
##  8    358    572    2225    597    4051
##  9    359      0      12      1       0
## 10    360    320     502     46    1114
## # … with 90 more rows
```

```r
# Tibbles carry and display extra information. While reading in it is easy to specify data type. 
untidy_counts &lt;- read_csv("dataset/hemato_rnaseq_counts.csv", col_types = cols(
    ENTREZ = col_character(),
    CD34_1 = col_integer(),
    ORTHO_1 = col_integer(),
    CD34_2 = col_integer(),
    ORTHO_2 = col_integer()
  ))
untidy_counts
```

```
## # A tibble: 100 x 5
##    ENTREZ CD34_1 ORTHO_1 CD34_2 ORTHO_2
##    &lt;chr&gt;   &lt;int&gt;   &lt;int&gt;  &lt;int&gt;   &lt;int&gt;
##  1 350       204       0    103       0
##  2 351     15586     479  10476      39
##  3 353       842     355   1188      86
##  4 354         0       0      0       0
##  5 355       123     291    139      16
##  6 356         1       1      0       0
##  7 357       380       3    177       0
##  8 358       572    2225    597    4051
##  9 359         0      12      1       0
## 10 360       320     502     46    1114
## # … with 90 more rows
```

# Tibbles

---
## Subsetting tibbles

```r
#Can use the same way as base to interact with the tibble dataframe
untidy_counts[,1]
```

```
## # A tibble: 100 x 1
##    ENTREZ
##    &lt;chr&gt; 
##  1 350   
##  2 351   
##  3 353   
##  4 354   
##  5 355   
##  6 356   
##  7 357   
##  8 358   
##  9 359   
## 10 360   
## # … with 90 more rows
```

```r
untidy_counts[1,]
```

```
## # A tibble: 1 x 5
##   ENTREZ CD34_1 ORTHO_1 CD34_2 ORTHO_2
##   &lt;chr&gt;   &lt;int&gt;   &lt;int&gt;  &lt;int&gt;   &lt;int&gt;
## 1 350       204       0    103       0
```

```r
#Can also not specify which dimension you pull from. This will default to grabbing the column
untidy_counts[1]
```

```
## # A tibble: 100 x 1
##    ENTREZ
##    &lt;chr&gt; 
##  1 350   
##  2 351   
##  3 353   
##  4 354   
##  5 355   
##  6 356   
##  7 357   
##  8 358   
##  9 359   
## 10 360   
## # … with 90 more rows
```

```r
#All the prior outputs have been outputting another tibble. If double brackets are used a vector is returned 
untidy_counts[[1]]
```

```
##   [1] "350" "351" "353" "354" "355" "356" "357" "358" "359" "360" "361"
##  [12] "362" "363" "364" "366" "367" "368" "369" "372" "373" "374" "375"
##  [23] "377" "378" "379" "381" "382" "383" "384" "387" "388" "389" "390"
##  [34] "391" "392" "393" "394" "395" "396" "397" "398" "399" "400" "401"
##  [45] "402" "403" "405" "406" "407" "408" "409" "410" "411" "412" "414"
##  [56] "415" "416" "417" "419" "420" "421" "427" "429" "430" "432" "433"
##  [67] "434" "435" "440" "443" "444" "445" "460" "462" "463" "466" "467"
##  [78] "468" "471" "472" "473" "474" "475" "476" "477" "478" "479" "480"
##  [89] "481" "482" "483" "486" "487" "488" "489" "490" "491" "492" "493"
## [100] "495"
```

```r
#This is also the case if you use the dollar and colname to access a column
untidy_counts$ENTREZ
```

```
##   [1] "350" "351" "353" "354" "355" "356" "357" "358" "359" "360" "361"
##  [12] "362" "363" "364" "366" "367" "368" "369" "372" "373" "374" "375"
##  [23] "377" "378" "379" "381" "382" "383" "384" "387" "388" "389" "390"
##  [34] "391" "392" "393" "394" "395" "396" "397" "398" "399" "400" "401"
##  [45] "402" "403" "405" "406" "407" "408" "409" "410" "411" "412" "414"
##  [56] "415" "416" "417" "419" "420" "421" "427" "429" "430" "432" "433"
##  [67] "434" "435" "440" "443" "444" "445" "460" "462" "463" "466" "467"
##  [78] "468" "471" "472" "473" "474" "475" "476" "477" "478" "479" "480"
##  [89] "481" "482" "483" "486" "487" "488" "489" "490" "491" "492" "493"
## [100] "495"
```

---
## CHALLENGE

* Read in count data as different data types. What happens when we extract the vector of these counts?

---
## ANSWER

```r
untidy_counts_test &lt;- read_csv("dataset/hemato_rnaseq_counts.csv", col_types = cols(
    ENTREZ = col_character(),
    CD34_1 = col_integer(),
    ORTHO_1 = col_character(),
    CD34_2 = col_factor(),
    ORTHO_2 = col_logical()
  ))
```

```
## Warning: 62 parsing failures.
## row     col           expected actual                               file
##   2 ORTHO_2 1/0/T/F/TRUE/FALSE   39   'dataset/hemato_rnaseq_counts.csv'
##   3 ORTHO_2 1/0/T/F/TRUE/FALSE   86   'dataset/hemato_rnaseq_counts.csv'
##   5 ORTHO_2 1/0/T/F/TRUE/FALSE   16   'dataset/hemato_rnaseq_counts.csv'
##   8 ORTHO_2 1/0/T/F/TRUE/FALSE   4051 'dataset/hemato_rnaseq_counts.csv'
##  10 ORTHO_2 1/0/T/F/TRUE/FALSE   1114 'dataset/hemato_rnaseq_counts.csv'
## ... ....... .................. ...... ..................................
## See problems(...) for more details.
```

```r
untidy_counts_test$CD34_1 %&gt;% head(n=8)
```

```
## [1]   204 15586   842     0   123     1   380   572
```

```r
untidy_counts_test$ORTHO_1 %&gt;% head(n=8)
```

```
## [1] "0"    "479"  "355"  "0"    "291"  "1"    "3"    "2225"
```

```r
untidy_counts_test$CD34_2 %&gt;% head(n=8)
```

```
## [1] 103   10476 1188  0     139   0     177   597  
## 82 Levels: 103 10476 1188 0 139 177 597 1 46 15 4 51 24 833 2741 ... 3053
```

```r
untidy_counts_test$ORTHO_2 %&gt;% head(n=8)
```

```
## [1] FALSE    NA    NA FALSE    NA FALSE FALSE    NA
```

---
## Converting Tibbles - Back and Forth

```r
# Can convert base dataframes into tibbles
as_tibble(untidy_counts_base)
```

```
## # A tibble: 100 x 5
##    ENTREZ CD34_1 ORTHO_1 CD34_2 ORTHO_2
##     &lt;int&gt;  &lt;int&gt;   &lt;int&gt;  &lt;int&gt;   &lt;int&gt;
##  1    350    204       0    103       0
##  2    351  15586     479  10476      39
##  3    353    842     355   1188      86
##  4    354      0       0      0       0
##  5    355    123     291    139      16
##  6    356      1       1      0       0
##  7    357    380       3    177       0
##  8    358    572    2225    597    4051
##  9    359      0      12      1       0
## 10    360    320     502     46    1114
## # … with 90 more rows
```

```r
# Once it is a tibble it is straight forward to modify the datatype
untidy_counts_base &lt;- as_tibble(untidy_counts_base) %&gt;%
  mutate_at(vars(ENTREZ), as.character)
untidy_counts_base
```

```
## # A tibble: 100 x 5
##    ENTREZ CD34_1 ORTHO_1 CD34_2 ORTHO_2
##    &lt;chr&gt;   &lt;int&gt;   &lt;int&gt;  &lt;int&gt;   &lt;int&gt;
##  1 350       204       0    103       0
##  2 351     15586     479  10476      39
##  3 353       842     355   1188      86
##  4 354         0       0      0       0
##  5 355       123     291    139      16
##  6 356         1       1      0       0
##  7 357       380       3    177       0
##  8 358       572    2225    597    4051
##  9 359         0      12      1       0
## 10 360       320     502     46    1114
## # … with 90 more rows
```

```r
# Some tools are not tibble friendly. Calling as.data.frame is sufficient to convert it back to a base data frame
as.data.frame(untidy_counts_base) %&gt;% head(n=12)
```

```
##    ENTREZ CD34_1 ORTHO_1 CD34_2 ORTHO_2
## 1     350    204       0    103       0
## 2     351  15586     479  10476      39
## 3     353    842     355   1188      86
## 4     354      0       0      0       0
## 5     355    123     291    139      16
## 6     356      1       1      0       0
## 7     357    380       3    177       0
## 8     358    572    2225    597    4051
## 9     359      0      12      1       0
## 10    360    320     502     46    1114
## 11    361      0       1      0       0
## 12    362      3       1     15       0
```

---
## Make your own tibble. Lets grab some metadata to do this. 

```r
# Lets load in some packages
library(org.Hs.eg.db)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)

# Lets use the ENTREZ ID as a key
keys &lt;- untidy_counts$ENTREZ

# We can use the ENTREZ ID to look up Gene Symbol
symbols &lt;- select(org.Hs.eg.db, keys = keys,columns = "SYMBOL", keytype = "ENTREZID")

# We can use the ENTREZ ID to look up the chormosome the gene resides on
chrs &lt;- select(TxDb.Hsapiens.UCSC.hg19.knownGene, keys = keys, columns = "TXCHROM", keytype = "GENEID")

# We can use the ENTREZ ID to get a list of genes with grange of their exons
geneExons &lt;- exonsBy(TxDb.Hsapiens.UCSC.hg19.knownGene,by = "gene")[keys]

# We will then use an apply to get the transcript length from each gene in the list. The transcript length is calculated by first flattening overalpping exons with reduce(), then calculating the length of each exon with width(), then summing upthe total exon length to get our transcript length. 
txsLength &lt;- sapply(geneExons, function(x){ x %&gt;%
    reduce() %&gt;%
    width() %&gt;%
    sum() })

# Finally we have all this metadata. Lets put it together into a tibble. 
counts_metadata &lt;- tibble(ID = symbols$ENTREZID, SYMBOL = symbols$SYMBOL, CHR = chrs$TXCHROM, LENGTH = txsLength)
```

---
## Tidying data up

What is wrong with the count dataframe from a tidy viewpoint?  

*Remember*. These are the rules:

* Each variable has its own column
* Each observation has its own row
* Each value has its own cell


```r
untidy_counts
```

```
## # A tibble: 100 x 5
##    ENTREZ CD34_1 ORTHO_1 CD34_2 ORTHO_2
##    &lt;chr&gt;   &lt;int&gt;   &lt;int&gt;  &lt;int&gt;   &lt;int&gt;
##  1 350       204       0    103       0
##  2 351     15586     479  10476      39
##  3 353       842     355   1188      86
##  4 354         0       0      0       0
##  5 355       123     291    139      16
##  6 356         1       1      0       0
##  7 357       380       3    177       0
##  8 358       572    2225    597    4051
##  9 359         0      12      1       0
## 10 360       320     502     46    1114
## # … with 90 more rows
```

A single variable with multiple columns

---
## How do we get tidy? - Pivot tools (formerly known as gather/spread )

```r
# Pivot longer allows you to collapse variables single varibles that are spread over multiple columns
tidier_counts &lt;- pivot_longer(untidy_counts, cols = c(-ENTREZ), names_to = c("Sample"), values_to = "counts")
tidier_counts
```

```
## # A tibble: 400 x 3
##    ENTREZ Sample  counts
##    &lt;chr&gt;  &lt;chr&gt;    &lt;int&gt;
##  1 350    CD34_1     204
##  2 350    ORTHO_1      0
##  3 350    CD34_2     103
##  4 350    ORTHO_2      0
##  5 351    CD34_1   15586
##  6 351    ORTHO_1    479
##  7 351    CD34_2   10476
##  8 351    ORTHO_2     39
##  9 353    CD34_1     842
## 10 353    ORTHO_1    355
## # … with 390 more rows
```

```r
# Pivot wider allows you to spread single variables over multiple columns   
pivot_wider(tidier_counts, names_from = c("Sample"), values_from = "counts")
```

```
## # A tibble: 100 x 5
##    ENTREZ CD34_1 ORTHO_1 CD34_2 ORTHO_2
##    &lt;chr&gt;   &lt;int&gt;   &lt;int&gt;  &lt;int&gt;   &lt;int&gt;
##  1 350       204       0    103       0
##  2 351     15586     479  10476      39
##  3 353       842     355   1188      86
##  4 354         0       0      0       0
##  5 355       123     291    139      16
##  6 356         1       1      0       0
##  7 357       380       3    177       0
##  8 358       572    2225    597    4051
##  9 359         0      12      1       0
## 10 360       320     502     46    1114
## # … with 90 more rows
```

---
## CHALLENGE

* tidy up df2

---
## ANSWER

```r
df2 %&gt;% pivot_wider(values_from = value, names_from = variable)
```

```
## # A tibble: 97 x 5
##    salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;
##  1     35032 Chinook salmon yearling                 147      41.3 
##  2     35033 Chinook salmon mixed age juvenile       444      62.1 
##  3     35034 Sockeye salmon juvenile                 139      NA   
##  4     35035 Sockeye salmon juvenile                 121      NA   
##  5     35036 Sockeye salmon juvenile                 112      NA   
##  6     35037 Steelhead      juvenile                 220      42.7 
##  7     35038 Steelhead      juvenile                 152      NA   
##  8     35039 Steelhead      juvenile                 201       9.92
##  9     35040 Steelhead      juvenile                 200      38.1 
## 10     35042 Steelhead      juvenile                 247      46.4 
## # … with 87 more rows
```

---
## What next?

```r
tidier_counts
```

```
## # A tibble: 400 x 3
##    ENTREZ Sample  counts
##    &lt;chr&gt;  &lt;chr&gt;    &lt;int&gt;
##  1 350    CD34_1     204
##  2 350    ORTHO_1      0
##  3 350    CD34_2     103
##  4 350    ORTHO_2      0
##  5 351    CD34_1   15586
##  6 351    ORTHO_1    479
##  7 351    CD34_2   10476
##  8 351    ORTHO_2     39
##  9 353    CD34_1     842
## 10 353    ORTHO_1    355
## # … with 390 more rows
```

Multiple variables in a single column

---
## Dealing with multiple variables in a single column


```r
#Seperate allows you to break a strings in a vraible by a seperator. In this case the cell type and replicate number are broken by underscore
tidier_counts &lt;- separate(tidier_counts, Sample, sep = "_", into=c("CellType","Rep"), remove=TRUE)
tidier_counts
```

```
## # A tibble: 400 x 4
##    ENTREZ CellType Rep   counts
##    &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;
##  1 350    CD34     1        204
##  2 350    ORTHO    1          0
##  3 350    CD34     2        103
##  4 350    ORTHO    2          0
##  5 351    CD34     1      15586
##  6 351    ORTHO    1        479
##  7 351    CD34     2      10476
##  8 351    ORTHO    2         39
##  9 353    CD34     1        842
## 10 353    ORTHO    1        355
## # … with 390 more rows
```

```r
#Unite can go the other way if you want to generate a key. We can use this to make sure we have a Key.
unite(tidier_counts, Sample, CellType, Rep, remove=FALSE)
```

```
## # A tibble: 400 x 5
##    ENTREZ Sample  CellType Rep   counts
##    &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;
##  1 350    CD34_1  CD34     1        204
##  2 350    ORTHO_1 ORTHO    1          0
##  3 350    CD34_2  CD34     2        103
##  4 350    ORTHO_2 ORTHO    2          0
##  5 351    CD34_1  CD34     1      15586
##  6 351    ORTHO_1 ORTHO    1        479
##  7 351    CD34_2  CD34     2      10476
##  8 351    ORTHO_2 ORTHO    2         39
##  9 353    CD34_1  CD34     1        842
## 10 353    ORTHO_1 ORTHO    1        355
## # … with 390 more rows
```

```r
#Remember you can always pipe eveyrthing together into a single expression
tidy_counts &lt;- untidy_counts %&gt;% 
  pivot_longer(cols = c(-ENTREZ), names_to = c("Sample"), values_to = "counts") %&gt;% 
  separate(Sample, sep = "_", into = c("CellType","Rep"), remove=FALSE)
```

# Joining: We have two related data frames. Can we link them together?

---
## Data frames can be joined on a shared variable *a.k.a.* a key
You need a key. We want this to be unique i.e. ENTREZ ID. 

```r
tidy_counts
```

```
## # A tibble: 400 x 5
##    ENTREZ Sample  CellType Rep   counts
##    &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;
##  1 350    CD34_1  CD34     1        204
##  2 350    ORTHO_1 ORTHO    1          0
##  3 350    CD34_2  CD34     2        103
##  4 350    ORTHO_2 ORTHO    2          0
##  5 351    CD34_1  CD34     1      15586
##  6 351    ORTHO_1 ORTHO    1        479
##  7 351    CD34_2  CD34     2      10476
##  8 351    ORTHO_2 ORTHO    2         39
##  9 353    CD34_1  CD34     1        842
## 10 353    ORTHO_1 ORTHO    1        355
## # … with 390 more rows
```

```r
counts_metadata
```

```
## # A tibble: 100 x 4
##    ID    SYMBOL  CHR   LENGTH
##    &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;  &lt;int&gt;
##  1 350   APOH    chr17   1201
##  2 351   APP     chr21   4480
##  3 353   APRT    chr16    807
##  4 354   KLK3    chr19   1906
##  5 355   FAS     chr10   6691
##  6 356   FASLG   chr1    1859
##  7 357   SHROOM2 chrX    8206
##  8 358   AQP1    chr7    3786
##  9 359   AQP2    chr12   4179
## 10 360   AQP3    chr9    2950
## # … with 90 more rows
```

```r
inner_join(tidy_counts, counts_metadata, by = c("ENTREZ" = "ID"))
```

```
## # A tibble: 400 x 8
##    ENTREZ Sample  CellType Rep   counts SYMBOL CHR   LENGTH
##    &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt;
##  1 350    CD34_1  CD34     1        204 APOH   chr17   1201
##  2 350    ORTHO_1 ORTHO    1          0 APOH   chr17   1201
##  3 350    CD34_2  CD34     2        103 APOH   chr17   1201
##  4 350    ORTHO_2 ORTHO    2          0 APOH   chr17   1201
##  5 351    CD34_1  CD34     1      15586 APP    chr21   4480
##  6 351    ORTHO_1 ORTHO    1        479 APP    chr21   4480
##  7 351    CD34_2  CD34     2      10476 APP    chr21   4480
##  8 351    ORTHO_2 ORTHO    2         39 APP    chr21   4480
##  9 353    CD34_1  CD34     1        842 APRT   chr16    807
## 10 353    ORTHO_1 ORTHO    1        355 APRT   chr16    807
## # … with 390 more rows
```

---
## There are many ways to join things

Inner Join 

* Keeps all observations in x and y with matching keys

Outer Join

* A left join keeps all observations in x and those in y with matching keys.
* A right join keeps all observations in y and those in x with matching keys. 
* A full join keeps all observations in x and y

---
## Example - I only want to look at expressed genes


```r
# In this pipe I group by gene, summarise the data based on the sum of counts, and filter for anything that has a count greater than 0. 
expressed_genes &lt;- tidy_counts %&gt;% 
  group_by(ENTREZ) %&gt;% 
  summarise(count_total=sum(counts)) %&gt;% 
  filter(count_total&gt;0)

# Left join shows all genes as my full data frame tidy_counts is used as the backbone. The filtered expressed genes is secondary, and has missing values (unexpressed genes) which are filled with NA
left_join(tidy_counts, expressed_genes, by = c("ENTREZ" = "ENTREZ")) %&gt;% print(n=20)
```

```
## # A tibble: 400 x 6
##    ENTREZ Sample  CellType Rep   counts count_total
##    &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;       &lt;int&gt;
##  1 350    CD34_1  CD34     1        204         307
##  2 350    ORTHO_1 ORTHO    1          0         307
##  3 350    CD34_2  CD34     2        103         307
##  4 350    ORTHO_2 ORTHO    2          0         307
##  5 351    CD34_1  CD34     1      15586       26580
##  6 351    ORTHO_1 ORTHO    1        479       26580
##  7 351    CD34_2  CD34     2      10476       26580
##  8 351    ORTHO_2 ORTHO    2         39       26580
##  9 353    CD34_1  CD34     1        842        2471
## 10 353    ORTHO_1 ORTHO    1        355        2471
## 11 353    CD34_2  CD34     2       1188        2471
## 12 353    ORTHO_2 ORTHO    2         86        2471
## 13 354    CD34_1  CD34     1          0          NA
## 14 354    ORTHO_1 ORTHO    1          0          NA
## 15 354    CD34_2  CD34     2          0          NA
## 16 354    ORTHO_2 ORTHO    2          0          NA
## 17 355    CD34_1  CD34     1        123         569
## 18 355    ORTHO_1 ORTHO    1        291         569
## 19 355    CD34_2  CD34     2        139         569
## 20 355    ORTHO_2 ORTHO    2         16         569
## # … with 380 more rows
```

```r
# Right join shows only genes that survived filtering as it is using the second dataframe as the backbone for the new dataframe. 
tidy_counts_expressed &lt;- right_join(tidy_counts,expressed_genes, by = c("ENTREZ" = "ENTREZ"))
tidy_counts_expressed  %&gt;% print(n=20)
```

```
## # A tibble: 376 x 6
##    ENTREZ Sample  CellType Rep   counts count_total
##    &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;       &lt;int&gt;
##  1 350    CD34_1  CD34     1        204         307
##  2 350    ORTHO_1 ORTHO    1          0         307
##  3 350    CD34_2  CD34     2        103         307
##  4 350    ORTHO_2 ORTHO    2          0         307
##  5 351    CD34_1  CD34     1      15586       26580
##  6 351    ORTHO_1 ORTHO    1        479       26580
##  7 351    CD34_2  CD34     2      10476       26580
##  8 351    ORTHO_2 ORTHO    2         39       26580
##  9 353    CD34_1  CD34     1        842        2471
## 10 353    ORTHO_1 ORTHO    1        355        2471
## 11 353    CD34_2  CD34     2       1188        2471
## 12 353    ORTHO_2 ORTHO    2         86        2471
## 13 355    CD34_1  CD34     1        123         569
## 14 355    ORTHO_1 ORTHO    1        291         569
## 15 355    CD34_2  CD34     2        139         569
## 16 355    ORTHO_2 ORTHO    2         16         569
## 17 356    CD34_1  CD34     1          1           2
## 18 356    ORTHO_1 ORTHO    1          1           2
## 19 356    CD34_2  CD34     2          0           2
## 20 356    ORTHO_2 ORTHO    2          0           2
## # … with 356 more rows
```

---
## Filtering joins. 

```r
#Semi join only keeps observations in x that are matched in y. y is only used as a reference and is not in output
semi_join(tidy_counts, expressed_genes)
```

```
## Joining, by = "ENTREZ"
```

```
## # A tibble: 376 x 5
##    ENTREZ Sample  CellType Rep   counts
##    &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;
##  1 350    CD34_1  CD34     1        204
##  2 350    ORTHO_1 ORTHO    1          0
##  3 350    CD34_2  CD34     2        103
##  4 350    ORTHO_2 ORTHO    2          0
##  5 351    CD34_1  CD34     1      15586
##  6 351    ORTHO_1 ORTHO    1        479
##  7 351    CD34_2  CD34     2      10476
##  8 351    ORTHO_2 ORTHO    2         39
##  9 353    CD34_1  CD34     1        842
## 10 353    ORTHO_1 ORTHO    1        355
## # … with 366 more rows
```

```r
#Anti join only keeps observations in x that are not matched in y. y is only used as a reference and is not in output
anti_join(tidy_counts, expressed_genes)
```

```
## Joining, by = "ENTREZ"
```

```
## # A tibble: 24 x 5
##    ENTREZ Sample  CellType Rep   counts
##    &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;
##  1 354    CD34_1  CD34     1          0
##  2 354    ORTHO_1 ORTHO    1          0
##  3 354    CD34_2  CD34     2          0
##  4 354    ORTHO_2 ORTHO    2          0
##  5 398    CD34_1  CD34     1          0
##  6 398    ORTHO_1 ORTHO    1          0
##  7 398    CD34_2  CD34     2          0
##  8 398    ORTHO_2 ORTHO    2          0
##  9 401    CD34_1  CD34     1          0
## 10 401    ORTHO_1 ORTHO    1          0
## # … with 14 more rows
```

---
## CHALLENGE
 
* Calculate CPMs and TPMs

-&gt; Show diagram of the math

---
## ANSWER

```r
# Use Group to focus computation on each sample. Use mutate to make new variable that is the CPM
tidy_counts_expressed &lt;- tidy_counts_expressed  %&gt;% 
  group_by(Sample) %&gt;% 
  mutate(CPM=(counts/sum(counts))*1000000)

# Join our tidy data to the metadata, then make new variable that is the TPM
tidy_counts_expressed &lt;- tidy_counts_expressed %&gt;% 
  inner_join(counts_metadata, by = c("ENTREZ" = "ID")) %&gt;%  
  mutate(TPM=(counts/(LENGTH/1000))/(sum(counts)/1000000))

# Simple X-Y plot comparing TPM and CPM. 
p &lt;- tidy_counts_expressed %&gt;% 
  ggplot(aes(x=CPM, y=TPM)) + 
  geom_point() + 
  scale_x_continuous(name="log2(CPM)",trans='log2') + 
  scale_y_continuous(name="log2(TPM)",trans='log2')
p
```

![](20191009_tidyr_files/figure-html/unnamed-chunk-30-1.png)&lt;!-- --&gt;

# Readr again: Writing your lovely new tibble to file

```r
# A key difference compared to base is that it does not write out row names. Tibbles generally don't have rownames. 
# Theres a wide range of writing options. Can specify the delmiter directly or use a specific function
write_delim(tidy_counts_expressed, '../expressed_genes_output.csv', delim =',')

write_csv(tidy_counts_expressed, '../expressed_genes_output.csv')
```

---
## At this point we have covered or touched on the most essential facets of tidy

* ~~ggplot2 – making pretty graphs~~ 
* ~~readr – reading data into R~~  
* ~~dplyr – manipulating data~~  
* ~~tibble - working with tibbles~~  
* ~~tidyr – miscellaneous tools for tidying data~~
* purrr - iterating over data  
* stringr – working with strings  
* forcats - working with factors  

# purrr - Functional programming 

* Applying functions to datasets  
* Base people use for loops or apply  
* Big advantage is that purrr readily handles nested dataframes and has standard outputs

---
## map - tidy way to iterate over a dataset 

```r
# Map is the tidy equivalent to apply. Here we take our untidy counts, trim of IDs, and then calculate means for each column. By default the output is a list
untidy_counts %&gt;% 
  dplyr::select(-ENTREZ) %&gt;% 
  map(mean)
```

```
## $CD34_1
## [1] 1497.67
## 
## $ORTHO_1
## [1] 822.33
## 
## $CD34_2
## [1] 1056.85
## 
## $ORTHO_2
## [1] 329.05
```

```r
# Same as the above line, but using map_dbl specifies the outputs is going to be a double
untidy_counts %&gt;% 
  dplyr::select(-ENTREZ) %&gt;% 
  map_dbl(mean)
```

```
##  CD34_1 ORTHO_1  CD34_2 ORTHO_2 
## 1497.67  822.33 1056.85  329.05
```

```r
# Summary also works in this context
tidy_counts %&gt;% 
  group_by(Sample) %&gt;% 
  summarize(mean_counts = mean(counts))
```

```
## # A tibble: 4 x 2
##   Sample  mean_counts
##   &lt;chr&gt;         &lt;dbl&gt;
## 1 CD34_1        1498.
## 2 CD34_2        1057.
## 3 ORTHO_1        822.
## 4 ORTHO_2        329.
```

```r
# This is an alternative method for doing this with an tidied frame
tidy_counts %&gt;% 
  split(.$Sample) %&gt;% 
  map_dbl(~mean(.$counts))
```

```
##  CD34_1  CD34_2 ORTHO_1 ORTHO_2 
## 1497.67 1056.85  822.33  329.05
```

```r
# pmap is a map variant for dealing with multiple inputs. This can be used to apply a function on a row by row basis
list(untidy_counts$ORTHO_1, untidy_counts$ORTHO_2) %&gt;% pmap_dbl(mean)
```

```
##   [1]     0   479   355     0   291     1     3  2225    12   502     1
##  [12]     1     6     0     0     0     0  1034  1864   728     5  6697
##  [23]   314  4845     1   574  2265  1632 10977  2457   171   193     0
##  [34]   577   304   138   464    67   702   410     0    11   147     0
##  [45]   218    25   824   175     1   111  1819   197   217    45    15
##  [56]     0     1     0     5  1136   255 11889     0    13     0     0
##  [67]     1   151  2520     0   151    15     0     0   118   480   402
##  [78]  6132   771   628   783     0    83  3453     3   121     0     5
##  [89]    18  1392  1454     0    33  1118   153  1263     1     0  3585
## [100]     0
```

---
## Nest - simplifying your dataframe by making it more complex

```r
# Nest all the data by sample
tidy_counts_nest &lt;- tidy_counts_expressed %&gt;% 
  group_by(Sample) %&gt;%
  nest()

#Looking at tibble it is a new datatype that appears simplified
tidy_counts_nest
```

```
## # A tibble: 4 x 2
## # Groups:   Sample [4]
##   Sample  data                                                             
##   &lt;chr&gt;   &lt;S3: vctrs_list_of&gt;                                              
## 1 CD34_1  350             , 351             , 353             , 355       …
## 2 ORTHO_1 350             , 351             , 353             , 355       …
## 3 CD34_2  350             , 351             , 353             , 355       …
## 4 ORTHO_2 350             , 351             , 353             , 355       …
```

```r
tidy_counts_nest$data %&gt;% is()
```

```
## [1] "vctrs_list_of"
```

```r
# The data is still there, nested within one of the variables
tidy_counts_nest$data[[1]]
```

```
## # A tibble: 94 x 10
##    ENTREZ CellType Rep   counts count_total    CPM SYMBOL CHR   LENGTH
##    &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;       &lt;int&gt;  &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt;
##  1 350    CD34     1        204         307 1.36e3 APOH   chr17   1201
##  2 351    CD34     1      15586       26580 1.04e5 APP    chr21   4480
##  3 353    CD34     1        842        2471 5.62e3 APRT   chr16    807
##  4 355    CD34     1        123         569 8.21e2 FAS    chr10   6691
##  5 356    CD34     1          1           2 6.68e0 FASLG  chr1    1859
##  6 357    CD34     1        380         560 2.54e3 SHROO… chrX    8206
##  7 358    CD34     1        572        7445 3.82e3 AQP1   chr7    3786
##  8 359    CD34     1          0          13 0.     AQP2   chr12   4179
##  9 360    CD34     1        320        1982 2.14e3 AQP3   chr9    2950
## 10 361    CD34     1          0           1 0.     AQP4   chr18   5217
## # … with 84 more rows, and 1 more variable: TPM &lt;dbl&gt;
```

```r
# Map can be used to apply functions across nested dataframes. Here we calculate a linear model. This is also saved in the tibble. 
tidy_counts_nest &lt;- tidy_counts_nest %&gt;% 
  mutate(my_model = map(data, ~lm(CPM ~ TPM, data = .)))

tidy_counts_nest
```

```
## # A tibble: 4 x 3
## # Groups:   Sample [4]
##   Sample  data                                                     my_model
##   &lt;chr&gt;   &lt;S3: vctrs_list_of&gt;                                      &lt;list&gt;  
## 1 CD34_1  350             , 351             , 353             , 3… &lt;S3: lm&gt;
## 2 ORTHO_1 350             , 351             , 353             , 3… &lt;S3: lm&gt;
## 3 CD34_2  350             , 351             , 353             , 3… &lt;S3: lm&gt;
## 4 ORTHO_2 350             , 351             , 353             , 3… &lt;S3: lm&gt;
```

```r
tidy_counts_nest$my_model %&gt;% is()
```

```
## [1] "list"             "vector"           "AssayData"       
## [4] "list_OR_List"     "vector_OR_factor"
```

```r
tidy_counts_nest$my_model[[1]]
```

```
## 
## Call:
## lm(formula = CPM ~ TPM, data = .)
## 
## Coefficients:
## (Intercept)          TPM  
##    3864.980        1.723
```

```r
# Tidy also has the ability to "tidy" up outputs from common statisitical packages, using broom.
library(broom)

tidy_counts_nest &lt;- tidy_counts_nest %&gt;% 
  mutate(my_tidy_model = map(my_model, broom::tidy))

tidy_counts_nest
```

```
## # A tibble: 4 x 4
## # Groups:   Sample [4]
##   Sample  data                                      my_model my_tidy_model 
##   &lt;chr&gt;   &lt;S3: vctrs_list_of&gt;                       &lt;list&gt;   &lt;list&gt;        
## 1 CD34_1  350             , 351             , 353 … &lt;S3: lm&gt; &lt;tibble [2 × …
## 2 ORTHO_1 350             , 351             , 353 … &lt;S3: lm&gt; &lt;tibble [2 × …
## 3 CD34_2  350             , 351             , 353 … &lt;S3: lm&gt; &lt;tibble [2 × …
## 4 ORTHO_2 350             , 351             , 353 … &lt;S3: lm&gt; &lt;tibble [2 × …
```

```r
tidy_counts_nest$my_tidy_model[[1]]
```

```
## # A tibble: 2 x 5
##   term        estimate std.error statistic  p.value
##   &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
## 1 (Intercept)  3865.    1101.         3.51 6.94e- 4
## 2 TPM             1.72     0.108     16.0  2.75e-28
```

---
## Unnest - Expand out dataframes

```r
#Unnesting to get everything back into a dataframe is very straightforward
tidy_counts_nest %&gt;%
  unnest(my_tidy_model)
```

```
## # A tibble: 8 x 8
## # Groups:   Sample [4]
##   Sample  data         my_model term  estimate std.error statistic  p.value
##   &lt;chr&gt;   &lt;S3: vctrs_&gt; &lt;list&gt;   &lt;chr&gt;    &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
## 1 CD34_1  350        … &lt;S3: lm&gt; (Int…  3865.   1101.          3.51 6.94e- 4
## 2 CD34_1  350        … &lt;S3: lm&gt; TPM       1.72    0.108      16.0  2.75e-28
## 3 ORTHO_1 350        … &lt;S3: lm&gt; (Int…  1522.    784.          1.94 5.53e- 2
## 4 ORTHO_1 350        … &lt;S3: lm&gt; TPM       2.20    0.0696     31.6  3.83e-51
## 5 CD34_2  350        … &lt;S3: lm&gt; (Int…  4083.   1085.          3.76 2.94e- 4
## 6 CD34_2  350        … &lt;S3: lm&gt; TPM       1.54    0.0944     16.3  7.03e-29
## 7 ORTHO_2 350        … &lt;S3: lm&gt; (Int…  1787.    937.          1.91 5.96e- 2
## 8 ORTHO_2 350        … &lt;S3: lm&gt; TPM       2.21    0.0884     25.0  8.45e-43
```

```r
#Unnesting can be done sequentially to keep adding to master dataframe
tidy_counts_nest %&gt;%
  unnest(my_tidy_model) %&gt;% 
  unnest(data)
```

```
## # A tibble: 752 x 17
## # Groups:   Sample [4]
##    Sample ENTREZ CellType Rep   counts count_total    CPM SYMBOL CHR  
##    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;       &lt;int&gt;  &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;
##  1 CD34_1 350    CD34     1        204         307 1.36e3 APOH   chr17
##  2 CD34_1 351    CD34     1      15586       26580 1.04e5 APP    chr21
##  3 CD34_1 353    CD34     1        842        2471 5.62e3 APRT   chr16
##  4 CD34_1 355    CD34     1        123         569 8.21e2 FAS    chr10
##  5 CD34_1 356    CD34     1          1           2 6.68e0 FASLG  chr1 
##  6 CD34_1 357    CD34     1        380         560 2.54e3 SHROO… chrX 
##  7 CD34_1 358    CD34     1        572        7445 3.82e3 AQP1   chr7 
##  8 CD34_1 359    CD34     1          0          13 0.     AQP2   chr12
##  9 CD34_1 360    CD34     1        320        1982 2.14e3 AQP3   chr9 
## 10 CD34_1 361    CD34     1          0           1 0.     AQP4   chr18
## # … with 742 more rows, and 8 more variables: LENGTH &lt;int&gt;, TPM &lt;dbl&gt;,
## #   my_model &lt;list&gt;, term &lt;chr&gt;, estimate &lt;dbl&gt;, std.error &lt;dbl&gt;,
## #   statistic &lt;dbl&gt;, p.value &lt;dbl&gt;
```

---
## CHALLENGE

---
## ANSWER


# stringr - working with strings

If the data you are working with involves characters from data entry often there will be errors i.e. clinical study metadata or a hand-typed list of genes of interest. Tidying data also means fixing these problems. stringr helps make this easy. 

* Access and manipulate characters
* Deal with whitspace
* Pattern Recognition

Though stringr is pretty comprehensive and covers most of what you will need, there is a sister package called stringi with even more functionality. 

---
## Basic functions to interact with strings
Many overlapping functions with base for combining, subsetting, converting and finding strings


```r
brc &lt;- c("Tom", "Ji-Dung", "Matt")

# Extract substrings from a range. Here the 1st to 3rd character
brc %&gt;% str_sub( 1, 3)
```

```
## [1] "Tom" "Ji-" "Mat"
```

```r
brc
```

```
## [1] "Tom"     "Ji-Dung" "Matt"
```

```r
# Extract substrings from a range. Here the 2nd to 2nd to last character
brc %&gt;% str_sub( 2, -2)
```

```
## [1] "o"     "i-Dun" "at"
```

```r
# Assign values back to substrings. Here the 2nd to 2nd to last character is replaced with X.
str_sub(brc, 2, -2)&lt;-'X'
brc
```

```
## [1] "TXm" "JXg" "MXt"
```

---
## Stripping whitespace

```r
brc2 &lt;- c("Tom  ", "  Ji  -Dung", "Matt   ")

#Trim whitespace from strings
brc2 &lt;- str_trim(brc2)
brc2 
```

```
## [1] "Tom"       "Ji  -Dung" "Matt"
```

```r
#Can add whitespace to strings to get consistent length. Here all are 10 characters
str_pad(brc2, width=10, side='left')
```

```
## [1] "       Tom" " Ji  -Dung" "      Matt"
```

---
## Converting strings - Capitalization

```r
#Lets reuse our counts tibble. pull from dplyr can be used to grab a tibble column and make it into a vector
tidy_counts_expressed %&gt;% 
  pull(SYMBOL) %&gt;%
  head()
```

```
## [1] "APOH" "APOH" "APOH" "APOH" "APP"  "APP"
```

```r
#Here we pull our gene symbols from our tibble into a vector, and then convert them into title style capitalization
tidy_counts_expressed %&gt;% 
  pull(SYMBOL) %&gt;% 
  str_to_title() %&gt;% 
  head()
```

```
## [1] "Apoh" "Apoh" "Apoh" "Apoh" "App"  "App"
```

```r
#String manipulation functions can be used on tibbles using mutate. Here we convert gene symbols to title style capitalization
tidy_counts_expressed %&gt;% 
  mutate(SYMBOL = str_to_title(SYMBOL))
```

```
## # A tibble: 376 x 11
## # Groups:   Sample [4]
##    ENTREZ Sample CellType Rep   counts count_total    CPM SYMBOL CHR  
##    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;       &lt;int&gt;  &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;
##  1 350    CD34_1 CD34     1        204         307 1.36e3 Apoh   chr17
##  2 350    ORTHO… ORTHO    1          0         307 0.     Apoh   chr17
##  3 350    CD34_2 CD34     2        103         307 9.75e2 Apoh   chr17
##  4 350    ORTHO… ORTHO    2          0         307 0.     Apoh   chr17
##  5 351    CD34_1 CD34     1      15586       26580 1.04e5 App    chr21
##  6 351    ORTHO… ORTHO    1        479       26580 5.82e3 App    chr21
##  7 351    CD34_2 CD34     2      10476       26580 9.91e4 App    chr21
##  8 351    ORTHO… ORTHO    2         39       26580 1.19e3 App    chr21
##  9 353    CD34_1 CD34     1        842        2471 5.62e3 Aprt   chr16
## 10 353    ORTHO… ORTHO    1        355        2471 4.32e3 Aprt   chr16
## # … with 366 more rows, and 2 more variables: LENGTH &lt;int&gt;, TPM &lt;dbl&gt;
```

```r
# Here we convert CHR annotation to capitals
tidy_counts_expressed %&gt;% 
  mutate(CHR = str_to_upper(CHR))
```

```
## # A tibble: 376 x 11
## # Groups:   Sample [4]
##    ENTREZ Sample CellType Rep   counts count_total    CPM SYMBOL CHR  
##    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;       &lt;int&gt;  &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;
##  1 350    CD34_1 CD34     1        204         307 1.36e3 APOH   CHR17
##  2 350    ORTHO… ORTHO    1          0         307 0.     APOH   CHR17
##  3 350    CD34_2 CD34     2        103         307 9.75e2 APOH   CHR17
##  4 350    ORTHO… ORTHO    2          0         307 0.     APOH   CHR17
##  5 351    CD34_1 CD34     1      15586       26580 1.04e5 APP    CHR21
##  6 351    ORTHO… ORTHO    1        479       26580 5.82e3 APP    CHR21
##  7 351    CD34_2 CD34     2      10476       26580 9.91e4 APP    CHR21
##  8 351    ORTHO… ORTHO    2         39       26580 1.19e3 APP    CHR21
##  9 353    CD34_1 CD34     1        842        2471 5.62e3 APRT   CHR16
## 10 353    ORTHO… ORTHO    1        355        2471 4.32e3 APRT   CHR16
## # … with 366 more rows, and 2 more variables: LENGTH &lt;int&gt;, TPM &lt;dbl&gt;
```

---
## Finding patterns

```r
#Find patterns in different ways
#Detect gives a T/F whether the pattern 'salmon' is present in vector
df1 %&gt;% 
  pull(common_name) %&gt;% 
  str_detect('salmon')
```

```
##  [1]  TRUE  TRUE  TRUE FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE FALSE
## [12] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
## [23] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE
## [34]  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE
## [45]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
## [56]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE
## [67] FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
## [78]  TRUE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
## [89]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
```

```r
#Subset returns the match if the pattern 'salmon' is present in vector
df1 %&gt;% 
  dplyr::pull(common_name) %&gt;% 
  str_subset('salmon') 
```

```
##  [1] "Chinook salmon" "Sockeye salmon" "Sockeye salmon" "Chinook salmon"
##  [5] "Sockeye salmon" "Chinook salmon" "Chinook salmon" "Sockeye salmon"
##  [9] "Chinook salmon" "Chinook salmon" "Sockeye salmon" "Chinook salmon"
## [13] "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon"
## [17] "Chinook salmon" "Chinook salmon" "Chinook salmon" "Coho salmon"   
## [21] "Coho salmon"    "Chinook salmon" "Chinook salmon" "Chinook salmon"
## [25] "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon"
## [29] "Sockeye salmon" "Sockeye salmon" "Sockeye salmon" "Sockeye salmon"
## [33] "Sockeye salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon"
## [37] "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon"
## [41] "Chinook salmon" "Chinook salmon" "Chinook salmon" "Sockeye salmon"
## [45] "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon"
## [49] "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon"
## [53] "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon"
## [57] "Chinook salmon" "Chinook salmon" "Chinook salmon"
```

```r
#Ends is similar to detect as it gives gives a T/F whether the pattern 'salmon' is present in vector, but the pattern has to be at the end. 
df1 %&gt;% 
  dplyr::pull(common_name) %&gt;% 
  str_ends('salmon') 
```

```
##  [1]  TRUE  TRUE  TRUE FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE FALSE
## [12] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
## [23] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE
## [34]  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE
## [45]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
## [56]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE
## [67] FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
## [78]  TRUE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
## [89]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
```

```r
df1 %&gt;% 
  filter(str_ends(common_name,'salmon'))
```

```
## # A tibble: 59 x 5
##    salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;
##  1     35032 Chinook salmon yearling                 147       41.3
##  2     35035 Sockeye salmon juvenile                 121       NA  
##  3     35036 Sockeye salmon juvenile                 112       NA  
##  4     35033 Chinook salmon mixed age juvenile       444       62.1
##  5     35034 Sockeye salmon juvenile                 139       NA  
##  6     35142 Chinook salmon yearling                 149       66.5
##  7     35143 Chinook salmon yearling                 204       80.9
##  8     35144 Sockeye salmon juvenile                 140       NA  
##  9     35145 Chinook salmon yearling                 130       23.4
## 10     35146 Chinook salmon mixed age juvenile       422      101. 
## # … with 49 more rows
```

```r
#Count gives you the total number of times your pattern appears in each chracter in the vector
df1 %&gt;% 
  dplyr::pull(common_name) %&gt;% 
  str_count('salmon')
```

```
##  [1] 1 1 1 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1
## [36] 1 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 0 0 1 1
## [71] 1 1 1 1 1 1 1 1 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
```

```r
df1 %&gt;% 
  dplyr::pull(common_name) %&gt;% 
  str_count('o')
```

```
##  [1] 3 2 2 0 0 3 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 3 3 2 3 3
## [36] 2 0 0 0 0 0 0 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 2 2 2 2 2 0 3 0 0 3 3
## [71] 3 3 3 3 3 3 3 2 0 0 0 0 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3
```

```r
#Replace
df1 %&gt;% 
  dplyr::pull(common_name) %&gt;%
  str_replace_all('Steelhead','Steelhead trout' )
```

```
##  [1] "Chinook salmon"  "Sockeye salmon"  "Sockeye salmon" 
##  [4] "Steelhead trout" "Steelhead trout" "Chinook salmon" 
##  [7] "Sockeye salmon"  "Steelhead trout" "Steelhead trout"
## [10] "Steelhead trout" "Steelhead trout" "Steelhead trout"
## [13] "Steelhead trout" "Steelhead trout" "Steelhead trout"
## [16] "Steelhead trout" "Steelhead trout" "Steelhead trout"
## [19] "Steelhead trout" "Steelhead trout" "Steelhead trout"
## [22] "Steelhead trout" "Steelhead trout" "Steelhead trout"
## [25] "Steelhead trout" "Steelhead trout" "Steelhead trout"
## [28] "Steelhead trout" "Steelhead trout" "Steelhead trout"
## [31] "Chinook salmon"  "Chinook salmon"  "Sockeye salmon" 
## [34] "Chinook salmon"  "Chinook salmon"  "Sockeye salmon" 
## [37] "Steelhead trout" "Steelhead trout" "Steelhead trout"
## [40] "Steelhead trout" "Steelhead trout" "Steelhead trout"
## [43] "Chinook salmon"  "Chinook salmon"  "Chinook salmon" 
## [46] "Chinook salmon"  "Chinook salmon"  "Chinook salmon" 
## [49] "Chinook salmon"  "Chinook salmon"  "Coho salmon"    
## [52] "Coho salmon"     "Chinook salmon"  "Chinook salmon" 
## [55] "Chinook salmon"  "Chinook salmon"  "Chinook salmon" 
## [58] "Chinook salmon"  "Chinook salmon"  "Sockeye salmon" 
## [61] "Sockeye salmon"  "Sockeye salmon"  "Sockeye salmon" 
## [64] "Sockeye salmon"  "Steelhead trout" "Chinook salmon" 
## [67] "Steelhead trout" "Steelhead trout" "Chinook salmon" 
## [70] "Chinook salmon"  "Chinook salmon"  "Chinook salmon" 
## [73] "Chinook salmon"  "Chinook salmon"  "Chinook salmon" 
## [76] "Chinook salmon"  "Chinook salmon"  "Sockeye salmon" 
## [79] "Steelhead trout" "Steelhead trout" "Steelhead trout"
## [82] "Steelhead trout" "Chinook salmon"  "Chinook salmon" 
## [85] "Chinook salmon"  "Chinook salmon"  "Chinook salmon" 
## [88] "Chinook salmon"  "Chinook salmon"  "Chinook salmon" 
## [91] "Chinook salmon"  "Chinook salmon"  "Chinook salmon" 
## [94] "Chinook salmon"  "Chinook salmon"  "Chinook salmon" 
## [97] "Chinook salmon"
```

```r
df1 %&gt;% 
  mutate(common_name = str_replace_all(common_name,'Steelhead','Steelhead trout' ))
```

```
## # A tibble: 97 x 5
##    salmon_id common_name     age_classbylength  length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;           &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;
##  1     35032 Chinook salmon  yearling                 147       41.3
##  2     35035 Sockeye salmon  juvenile                 121       NA  
##  3     35036 Sockeye salmon  juvenile                 112       NA  
##  4     35037 Steelhead trout juvenile                 220       42.7
##  5     35038 Steelhead trout juvenile                 152       NA  
##  6     35033 Chinook salmon  mixed age juvenile       444       62.1
##  7     35034 Sockeye salmon  juvenile                 139       NA  
##  8     35048 Steelhead trout juvenile                 288       24.2
##  9     35049 Steelhead trout juvenile                 190       NA  
## 10     35050 Steelhead trout juvenile                 283       63.5
## # … with 87 more rows
```

---
## CHALLENGE

---
## ANSWER


# forcats - Handling factors

Factors are a data type that R uses to handle fixed categorical variables that have a known set of possible values. 

Factors are ordered, allowing hierachy to be presevred in relatively simple vectors.

---
## Making a factor - This is all base

```r
# Vectors are easy to turn into factors with factor()
tidy_counts_expressed_samples &lt;- tidy_counts_expressed %&gt;% 
  pull(Sample) %&gt;% 
  factor() 

tidy_counts_expressed_samples %&gt;% head(n=10)
```

```
##  [1] CD34_1  ORTHO_1 CD34_2  ORTHO_2 CD34_1  ORTHO_1 CD34_2  ORTHO_2
##  [9] CD34_1  ORTHO_1
## Levels: CD34_1 CD34_2 ORTHO_1 ORTHO_2
```

```r
# Can also modify the data type of a tibble column with as_facotr, in an approach we have used before.
tidy_counts_expressed %&gt;% 
  ungroup() %&gt;% 
  mutate_at(vars(Sample), as_factor) 
```

```
## # A tibble: 376 x 11
##    ENTREZ Sample CellType Rep   counts count_total    CPM SYMBOL CHR  
##    &lt;chr&gt;  &lt;fct&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;       &lt;int&gt;  &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;
##  1 350    CD34_1 CD34     1        204         307 1.36e3 APOH   chr17
##  2 350    ORTHO… ORTHO    1          0         307 0.     APOH   chr17
##  3 350    CD34_2 CD34     2        103         307 9.75e2 APOH   chr17
##  4 350    ORTHO… ORTHO    2          0         307 0.     APOH   chr17
##  5 351    CD34_1 CD34     1      15586       26580 1.04e5 APP    chr21
##  6 351    ORTHO… ORTHO    1        479       26580 5.82e3 APP    chr21
##  7 351    CD34_2 CD34     2      10476       26580 9.91e4 APP    chr21
##  8 351    ORTHO… ORTHO    2         39       26580 1.19e3 APP    chr21
##  9 353    CD34_1 CD34     1        842        2471 5.62e3 APRT   chr16
## 10 353    ORTHO… ORTHO    1        355        2471 4.32e3 APRT   chr16
## # … with 366 more rows, and 2 more variables: LENGTH &lt;int&gt;, TPM &lt;dbl&gt;
```

```r
tidy_counts_expressed %&gt;% 
  ungroup() %&gt;% 
  mutate_at(vars(Sample), as_factor) %&gt;% 
  pull(Sample) %&gt;% head(n=10)
```

```
##  [1] CD34_1  ORTHO_1 CD34_2  ORTHO_2 CD34_1  ORTHO_1 CD34_2  ORTHO_2
##  [9] CD34_1  ORTHO_1
## Levels: CD34_1 ORTHO_1 CD34_2 ORTHO_2
```

```r
# When you factorize you can use a vector to determine the order
my_levels1&lt;-c('ORTHO_1','ORTHO_2','CD34_1','CD34_2')
tidy_counts_expressed %&gt;% 
  pull(Sample) %&gt;% 
  factor(levels = my_levels1 ) %&gt;% head(n=10)
```

```
##  [1] CD34_1  ORTHO_1 CD34_2  ORTHO_2 CD34_1  ORTHO_1 CD34_2  ORTHO_2
##  [9] CD34_1  ORTHO_1
## Levels: ORTHO_1 ORTHO_2 CD34_1 CD34_2
```

```r
# When you factorize anything not in the given levels is turned to NA
my_levels2&lt;-c('ORTHO_1','CD34_1')
tidy_counts_expressed %&gt;% 
  pull(Sample) %&gt;% 
  factor(levels = my_levels2 ) %&gt;% head(n=10)
```

```
##  [1] CD34_1  ORTHO_1 &lt;NA&gt;    &lt;NA&gt;    CD34_1  ORTHO_1 &lt;NA&gt;    &lt;NA&gt;   
##  [9] CD34_1  ORTHO_1
## Levels: ORTHO_1 CD34_1
```

```r
# Its straightforward to grab the levels from the factor
tidy_counts_expressed_samples %&gt;% 
  levels()
```

```
## [1] "CD34_1"  "CD34_2"  "ORTHO_1" "ORTHO_2"
```

---
## Why do we factorize? For order:

```r
tidy_counts_expressed %&gt;% 
  ungroup() %&gt;% 
  mutate_at(vars(Sample), as_factor) %&gt;% 
  group_by(Sample) %&gt;% 
  summarize(mean_count=mean(counts)) %&gt;% 
  ggplot(aes(x=Sample, y= mean_count)) + 
  geom_bar(stat = "identity") + 
  theme(axis.text.x = element_text(angle = 90))
```

![](20191009_tidyr_files/figure-html/unnamed-chunk-42-1.png)&lt;!-- --&gt;

---
## Changing the order

```r
#fct_relevel - reorder manually
tidy_counts_expressed %&gt;% 
  ungroup() %&gt;% 
  mutate_at(vars(Sample), as_factor) %&gt;%
  mutate(Sample = fct_relevel(Sample, my_levels1)) %&gt;% 
  pull(Sample) %&gt;% head(n=10)
```

```
##  [1] CD34_1  ORTHO_1 CD34_2  ORTHO_2 CD34_1  ORTHO_1 CD34_2  ORTHO_2
##  [9] CD34_1  ORTHO_1
## Levels: ORTHO_1 ORTHO_2 CD34_1 CD34_2
```

```r
#fct_relevel - reorder manually
tidy_counts_expressed %&gt;% 
  ungroup() %&gt;% 
  mutate_at(vars(Sample), as_factor) %&gt;%
  mutate(Sample = fct_relevel(Sample, my_levels2)) %&gt;% 
  pull(Sample) %&gt;% head(n=10)
```

```
##  [1] CD34_1  ORTHO_1 CD34_2  ORTHO_2 CD34_1  ORTHO_1 CD34_2  ORTHO_2
##  [9] CD34_1  ORTHO_1
## Levels: ORTHO_1 CD34_1 CD34_2 ORTHO_2
```

```r
#fct_reorder - reorder based on the data. Here we are ordering based on mean counts for each sample.
tidy_counts_expressed %&gt;%
  ungroup() %&gt;% 
  mutate_at(vars(Sample), as_factor) %&gt;% 
  mutate(Sample = fct_reorder(Sample, counts, mean)) %&gt;% 
  pull(Sample) %&gt;% head(n=10)
```

```
##  [1] CD34_1  ORTHO_1 CD34_2  ORTHO_2 CD34_1  ORTHO_1 CD34_2  ORTHO_2
##  [9] CD34_1  ORTHO_1
## Levels: ORTHO_2 ORTHO_1 CD34_2 CD34_1
```

---
## Now our ordering can be fixed

```r
tidy_counts_expressed %&gt;% 
  ungroup() %&gt;% 
  mutate_at(vars(Sample), as_factor) %&gt;%
  mutate(Sample=fct_relevel(Sample, my_levels1)) %&gt;% 
  group_by(Sample) %&gt;% 
  summarize(mean_count=mean(counts)) %&gt;% 
  ggplot(aes(x=Sample, y= mean_count)) + 
  geom_bar(stat = "identity") + 
  theme(axis.text.x = element_text(angle = 90))
```

![](20191009_tidyr_files/figure-html/unnamed-chunk-44-1.png)&lt;!-- --&gt;

---
## Other useful things.

```r
df1$age_classbylength
```

```
##  [1] "yearling"           "juvenile"           "juvenile"          
##  [4] "juvenile"           "juvenile"           "mixed age juvenile"
##  [7] "juvenile"           "juvenile"           "juvenile"          
## [10] "juvenile"           "juvenile"           "juvenile"          
## [13] "juvenile"           "juvenile"           "juvenile"          
## [16] "juvenile"           "juvenile"           "juvenile"          
## [19] "juvenile"           "juvenile"           "juvenile"          
## [22] "juvenile"           "juvenile"           "juvenile"          
## [25] "juvenile"           "juvenile"           "juvenile"          
## [28] "juvenile"           "juvenile"           "juvenile"          
## [31] "yearling"           "yearling"           "juvenile"          
## [34] "yearling"           "mixed age juvenile" "juvenile"          
## [37] "juvenile"           "juvenile"           "juvenile"          
## [40] "juvenile"           "juvenile"           "juvenile"          
## [43] "yearling"           "yearling"           "yearling"          
## [46] "yearling"           "yearling"           "yearling"          
## [49] "yearling"           "yearling"           "yearling"          
## [52] "yearling"           "yearling"           "yearling"          
## [55] "yearling"           "yearling"           "yearling"          
## [58] "yearling"           "subyearling"        "juvenile"          
## [61] "juvenile"           "juvenile"           "juvenile"          
## [64] "juvenile"           "juvenile"           "yearling"          
## [67] "juvenile"           "juvenile"           "mixed age juvenile"
## [70] "yearling"           "yearling"           "yearling"          
## [73] "yearling"           "yearling"           "yearling"          
## [76] "yearling"           "yearling"           "juvenile"          
## [79] "juvenile"           "juvenile"           "juvenile"          
## [82] "juvenile"           "yearling"           "yearling"          
## [85] "yearling"           "yearling"           "yearling"          
## [88] "yearling"           "yearling"           "yearling"          
## [91] "yearling"           "yearling"           "yearling"          
## [94] "yearling"           "yearling"           "yearling"          
## [97] "yearling"
```

```r
# fct_recode - Change levels manually
df1 %&gt;% mutate_at(vars(age_classbylength), as_factor) %&gt;% mutate(age_classbylength=fct_recode(age_classbylength, "mixed juvenile" = "mixed age juvenile")) %&gt;% pull(age_classbylength)
```

```
##  [1] yearling       juvenile       juvenile       juvenile      
##  [5] juvenile       mixed juvenile juvenile       juvenile      
##  [9] juvenile       juvenile       juvenile       juvenile      
## [13] juvenile       juvenile       juvenile       juvenile      
## [17] juvenile       juvenile       juvenile       juvenile      
## [21] juvenile       juvenile       juvenile       juvenile      
## [25] juvenile       juvenile       juvenile       juvenile      
## [29] juvenile       juvenile       yearling       yearling      
## [33] juvenile       yearling       mixed juvenile juvenile      
## [37] juvenile       juvenile       juvenile       juvenile      
## [41] juvenile       juvenile       yearling       yearling      
## [45] yearling       yearling       yearling       yearling      
## [49] yearling       yearling       yearling       yearling      
## [53] yearling       yearling       yearling       yearling      
## [57] yearling       yearling       subyearling    juvenile      
## [61] juvenile       juvenile       juvenile       juvenile      
## [65] juvenile       yearling       juvenile       juvenile      
## [69] mixed juvenile yearling       yearling       yearling      
## [73] yearling       yearling       yearling       yearling      
## [77] yearling       juvenile       juvenile       juvenile      
## [81] juvenile       juvenile       yearling       yearling      
## [85] yearling       yearling       yearling       yearling      
## [89] yearling       yearling       yearling       yearling      
## [93] yearling       yearling       yearling       yearling      
## [97] yearling      
## Levels: yearling juvenile mixed juvenile subyearling
```

```r
df1 %&gt;% mutate_at(vars(age_classbylength), as_factor) %&gt;% mutate(age_classbylength=fct_recode(age_classbylength, "juvenile" = "mixed age juvenile")) %&gt;% pull(age_classbylength)
```

```
##  [1] yearling    juvenile    juvenile    juvenile    juvenile   
##  [6] juvenile    juvenile    juvenile    juvenile    juvenile   
## [11] juvenile    juvenile    juvenile    juvenile    juvenile   
## [16] juvenile    juvenile    juvenile    juvenile    juvenile   
## [21] juvenile    juvenile    juvenile    juvenile    juvenile   
## [26] juvenile    juvenile    juvenile    juvenile    juvenile   
## [31] yearling    yearling    juvenile    yearling    juvenile   
## [36] juvenile    juvenile    juvenile    juvenile    juvenile   
## [41] juvenile    juvenile    yearling    yearling    yearling   
## [46] yearling    yearling    yearling    yearling    yearling   
## [51] yearling    yearling    yearling    yearling    yearling   
## [56] yearling    yearling    yearling    subyearling juvenile   
## [61] juvenile    juvenile    juvenile    juvenile    juvenile   
## [66] yearling    juvenile    juvenile    juvenile    yearling   
## [71] yearling    yearling    yearling    yearling    yearling   
## [76] yearling    yearling    juvenile    juvenile    juvenile   
## [81] juvenile    juvenile    yearling    yearling    yearling   
## [86] yearling    yearling    yearling    yearling    yearling   
## [91] yearling    yearling    yearling    yearling    yearling   
## [96] yearling    yearling   
## Levels: yearling juvenile subyearling
```

```r
# fct_count - Count up the number of each
df1 %&gt;% mutate_at(vars(age_classbylength), as_factor) %&gt;% pull(age_classbylength) %&gt;% fct_count()
```

```
## # A tibble: 4 x 2
##   f                      n
##   &lt;fct&gt;              &lt;int&gt;
## 1 yearling              44
## 2 juvenile              49
## 3 mixed age juvenile     3
## 4 subyearling            1
```

```r
# fct_infreq - mask rare factors
df1 %&gt;% mutate_at(vars(age_classbylength), as_factor) %&gt;% mutate(age_classbylength=fct_lump(age_classbylength, n=2)) %&gt;% pull(age_classbylength)
```

```
##  [1] yearling juvenile juvenile juvenile juvenile Other    juvenile
##  [8] juvenile juvenile juvenile juvenile juvenile juvenile juvenile
## [15] juvenile juvenile juvenile juvenile juvenile juvenile juvenile
## [22] juvenile juvenile juvenile juvenile juvenile juvenile juvenile
## [29] juvenile juvenile yearling yearling juvenile yearling Other   
## [36] juvenile juvenile juvenile juvenile juvenile juvenile juvenile
## [43] yearling yearling yearling yearling yearling yearling yearling
## [50] yearling yearling yearling yearling yearling yearling yearling
## [57] yearling yearling Other    juvenile juvenile juvenile juvenile
## [64] juvenile juvenile yearling juvenile juvenile Other    yearling
## [71] yearling yearling yearling yearling yearling yearling yearling
## [78] juvenile juvenile juvenile juvenile juvenile yearling yearling
## [85] yearling yearling yearling yearling yearling yearling yearling
## [92] yearling yearling yearling yearling yearling yearling
## Levels: yearling juvenile Other
```

```r
df1 %&gt;% mutate_at(vars(age_classbylength), as_factor) %&gt;% mutate(age_classbylength=fct_lump(age_classbylength, n=2)) %&gt;% pull(age_classbylength) %&gt;% fct_count()
```

```
## # A tibble: 3 x 2
##   f            n
##   &lt;fct&gt;    &lt;int&gt;
## 1 yearling    44
## 2 juvenile    49
## 3 Other        4
```

```r
# fct_c - combining factors
A&lt;-factor(c('Tom','Ji-Dung'))
B&lt;-factor('Matt')
fct_c(A,B)
```

```
## [1] Tom     Ji-Dung Matt   
## Levels: Ji-Dung Tom Matt
```

---
## CHALLENGE
short/long genes
---
## ANSWER


# Summary
---
## Tidy packages we have covered

* ~~ggplot2 – making pretty graphs~~ 
* ~~readr – reading data into R~~  
* ~~dplyr – manipulating data~~  
* ~~tibble - working with tibbles~~  
* ~~tidyr – miscellaneous tools for tidying data~~
* ~~purrr - iterating over data~~
* ~~stringr – working with strings~~  
* ~~forcats - working with factors~~  

---
## Tidy beyond this workshop

Hadley Wickham (Chief Scientist at RStudio) is the driving force behind the tidyverse.  

Hadley wrote a paper about why he thinks tidy data is best: www.jstatsoft.org/v59/i10/paper.  

There is a lot of support for all things tidy at: https://www.tidyverse.org/  
(This includes really great cheat sheets for each tool)

---
## Tidy packages to check out:

plyranges: dplyr equivalent for working with ranges i.e. Granges

readxl: This package is very useful when you want to import Excel sheets in R

googledrive: Interact with your googledrive through R

lubridate and hms: Allow managin of calendar and time formats

magrittr: piping tools

broom: helps tidy up standard base function i.e. lm or t.test

tidymodels: A collection of tools for preparing for and validating model functions

---
## Other Good Resources

ggplot  
https://rockefelleruniversity.github.io/Plotting_In_R/

R for Data Science text book  
https://r4ds.had.co.nz/

Data Science with R text book
http://garrettgman.github.io/

Cancer Research UK - Intermediate R course
https://bioinformatics-core-shared-training.github.io/r-intermediate/

# Thanks

*BRC*
IT Pavillion
Tom Carroll: tcarroll@rockefeller.edu
Ji-Dung Luo: jluo@rockefeller.edu
Matt Paul: mpaul@rockefeller.edu

*Data Science Research Group*
Eli:
Doug:
Greg:
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
